functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java

```diff
@@ -0,0 +1,134 @@
+import software.amazon.lambda.powertools.idempotency.persistence.DynamoDBPersistenceStore;
+import software.amazon.lambda.powertools.logging.Logging;
+import software.amazon.lambda.powertools.logging.LoggingUtils;
+import software.amazon.lambda.powertools.metrics.Metrics;
+import software.amazon.lambda.powertools.tracing.Tracing;
+
+import static com.amazonaws.powertools.workshop.Utils.getImageMetadata;
+import static com.amazonaws.powertools.workshop.Utils.markFileAs;
+import static software.amazon.lambda.powertools.metrics.MetricsUtils.metricsLogger;
+
+/**
+ * Handler for requests to Lambda function.
+ */
+public class Module1Handler implements RequestHandler<S3EBEvent, String> {
+    private static final String IDEMPOTENCY_TABLE_NAME = System.getenv("IDEMPOTENCY_TABLE_NAME");
+
+    private static final String  TRANSFORMED_IMAGE_PREFIX = "transformed/image/jpg";
+    private static final String  TRANSFORMED_IMAGE_EXTENSION = ".jpeg";
+    private static final int TRANSFORMED_IMAGE_WIDTH = 720;
+    private static final int TRANSFORMED_IMAGE_HEIGHT = 480;
+    private static final Logger LOGGER = LogManager.getLogger(Module1Handler.class);
+
+    public Module1Handler() {
+        // Configure Idempotency module
+        Idempotency.config().withConfig(
+                        IdempotencyConfig.builder()
+                                .withEventKeyJMESPath("etag") // TODO: replace with [etag, userId] when #1419 is fixed
+                                .withThrowOnNoIdempotencyKey(true)
+                                .withExpiration(Duration.ofMinutes(120))
+                                .build())
+                .withPersistenceStore(
+                        DynamoDBPersistenceStore.builder()
+                                .withDynamoDbClient(Utils.ddbClient)
+                                .withTableName(IDEMPOTENCY_TABLE_NAME)
+                                .build()
+                ).configure();
+    }
+
+    @Logging(logEvent = true)
+    @Tracing
+    @Metrics(captureColdStart = true)
+    public String handleRequest(final S3EBEvent event, final Context context) {
+        Idempotency.registerLambdaContext(context);
+
+        S3Object object = new S3Object();
+        object.setKey(event.getDetail().getObject().get("key"));
+        object.setEtag(event.getDetail().getObject().get("etag"));
+
+        // Fetch additional metadata from DynamoDB
+        Map<String, String> metadata = getImageMetadata(object.getKey());
+        object.setFileId(metadata.get("fileId"));
+        object.setUserId(metadata.get("userId"));
+
+        // add metadata to the logs
+        LoggingUtils.appendKey("S3Object", object.toString());
+
+        try {
+            // Mark file as working
+            markFileAs(object.getFileId(), "in-progress", null);
+
+            String newObjectKey = processOneIdempotently(object);
+
+            // Mark file as done
+            markFileAs(object.getFileId(), "completed", newObjectKey);
```

Comment:
If this fails, we'll flip back to `markFileAs(..., "in-progress", ...)` and then hit the idemp cache on :82 . is that fine? Could we also do all of this in the idempotent method?

Reply:
this is how they do in typescript, I keep the logic